textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/regexmatch/RegexInfo.java

```diff
@@ -78,5 +78,100 @@ static RegexInfo anyChar() {
 		regexInfo.emptyable = false;
 		return regexInfo;
 	}
+	
+	/**
+	 * This function simplifies the regexInfo. <br>
+	 * If the "exact" set gets too large, it will add exact to the query tree,
+	 * and move relevant pieces into prefix and suffix. <br>
+	 * 
+	 * Simplification is performed under three circumstances. <br>
+	 * 1, the size ofexact set is larger than a threshold. <br>
+	 * 2, the minimum length of strings in exact is larger than {@code gramLength+1}, 
+	 *    if we do not set {@code force = true}. <br>
+	 * 3, the minimum length of strings in exact is larger than {@code gramLength}, 
+	 *    if we set {@code force = true}. <br>
+	 * @param force
+	 */
+	RegexInfo simplify(boolean force) {
+		TranslatorUtils.removeDuplicateAffix(exact, false);
+		int gramLength = this.match.gramLength;
+		
+		if ( exact.size() > TranslatorUtils.MAX_EXACT_SIZE ||
+			( TranslatorUtils.minLenOfString(exact) >= gramLength && force) ||
+			TranslatorUtils.minLenOfString(exact) >= gramLength + 1){
+			// Add exact to match (query tree)
+			// Transfer information from exact to prefix and suffix
+			match.add(exact);
+			for (String str: exact) {
+				if (str.length() < gramLength) {
+					prefix.add(str);
+					suffix.add(str);
+				} else {
+					prefix.add(str.substring(0, gramLength));
+					suffix.add(str.substring(str.length()-gramLength, str.length()));
+				}
+			}
+			exact.clear();
+		}
+		
+		// Add information in prefix and suffix to match
+		if (exact.isEmpty()) {
+			simplifyAffix(prefix, false);
+			simplifyAffix(suffix, true);
+		}
+		
+		return this;
+	}
+	
+	/**
+	 * simplifyAffix reduces the size of the given set (either prefix or suffix).
+	 * If the set gets too big, it moves information in prefix/suffix into match query.
+	 * @param strList
+	 * @param isSuffix indicates if given string list is suffix list or not
+	 */
+	void simplifyAffix(List<String> strList, boolean isSuffix) {
+		TranslatorUtils.removeDuplicateAffix(strList, isSuffix);
+		int gramLength = this.match.gramLength;
+		
+		// Add the current prefix/suffix set to match query.
+		match.add(strList);
+		
+		// This loop cuts the length of prefix/suffix. It cuts all
+		// strings longer than {@code gramLength}, and continues to cut strings
```

Comment:
When strings in "prefix" ("suffix") become to long, we simplify "prefix" ("suffix") by only retaining the first (last) n characters.
For example, prefix: "abcd" ----> "abc". suffix: "abcd" ----> "bcd".
We throw some characters away to keep strings short.